pisek/config/did_you_mean.py

```diff
@@ -0,0 +1,190 @@
+
+
+class ConfigKeysHelper:
+    def __init__(self) -> None:
+        self.sections: dict[str, ConfigDescriptionSection] = {}
+        self.keys: dict[tuple[str, str], ConfigDescriptionKey] = {}
+        add_applicability_conditions: list[
+            tuple[ConfigDescriptionKey, Callable[[], ApplicabilityCondition]]
+        ] = []
+        with open(CONFIG_DESCRIPTION) as f:
+            section: Optional[ConfigDescriptionSection] = None
+            last_key: Optional[ConfigDescriptionKey] = None
+            for line in f:
+                line = line.strip()
+
+                if len(line) == 0:
+                    pass
+
+                elif line.startswith("#!"):
+                    assert section is not None
+                    [fun, *args] = line.removeprefix("#!").split()
+                    if fun == "regex":
+                        assert last_key is None
+                        section.similarity_function = regex_cmp
+                    elif fun == "if":
+                        assert section is not None
+                        assert last_key is not None
+                        if len(args) != 2 or args[1].count("=") != 1:
+                            self._invalid_function_args(fun, args)
+                        key_name, value = args[1].split("=")
+                        add_applicability_conditions.append(
+                            (last_key, self._gen_has_keyvalue(args[0], key_name, value))
+                        )
+                    elif fun == "default":
+                        if last_key is None:
+                            if len(args) != 1:
+                                self._invalid_function_args(fun, args)
+                            section.defaults_to.append(args[0])
+                        else:
+                            if len(args) != 2:
+                                self._invalid_function_args(fun, args)
+                            last_key.defaults_to.append((args[0], args[1]))
+                    else:
+                        raise ValueError(
+                            f"invalid config-description function: '{fun}'"
+                        )
+
+                elif line.count("=") == 1 and line[-1] == "=":
+                    assert section is not None
+                    last_key = ConfigDescriptionKey(section, line.removesuffix("="))
+                    self.keys[(section.section, last_key.key)] = last_key
+
+                elif line[0] == "[" and line[-1] == "]":
+                    last_key = None
+                    section = ConfigDescriptionSection(
+                        line.removeprefix("[").removesuffix("]")
+                    )
+                    self.sections[section.section] = section
+
+                else:
+                    raise ValueError(f"invalid config-description line: '{line}'")
+
+        for key, lambda_cond in add_applicability_conditions:
+            key.applicability_conditions.append(lambda_cond())
```

Comment:
Why is this deferred until the end of parsing?

Reply:
Because this key might not have been loaded yet